### ISSUE TITLE
LSDMap stage fails on stampede -- possibly due to compiler mismatch during mpirun_sh

### ISSUE THREAD

**Issue by vivek-bala** [radical-cybertools/radical.pilot#572]
LSDMap is installed using the intel/14* compiler (and cannot be run with intel/13* due to some specific python error). RP executes mpirun_sh from the intel/13* compiler.

The shell script looks like,

```
# Change to working directory for unit
cd /work/02734/vivek91/radical.pilot.sandbox/rp.session.ip-10-184-31-85.vivek.016546.0000-pilot.0000/unit.000017
# Pre-exec commands
module load gromacs
python pre_analyze.py 16 tmp.gro
echo 2 | trjconv -f tmp.gro -s tmp.gro -o tmpha.gro
module load -intel +intel/14.0.1.106
export PYTHONPATH=/home1/03036/jp43/.local/lib/python2.7/site-packages
module load python/2.7.6
export PATH=/home1/03036/jp43/.local/bin:$PATH
# The command to run
/opt/apps/intel13/mvapich2/1.9/bin/mpirun_rsh -np 16 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 c558-802 LD_LIBRARY_PATH=$LD_LIBRARY_PATH PATH=$PATH PYTHONPATH=$PYTHONPATH /opt/apps/intel14/mvapich2_2_0/python/2.7.6/lib/python2.7/site-packages/mpi4py/bin/python-mpi "lsdm.py" "-f" "config.ini" "-c" "tmpha.gro" "-n" "neighbors.nn" "-w" "weight.w"
```
The corresponding STDERR,

```
...
...
..
The following have been reloaded with a version change:
  1) intel/13.0.2.146 => intel/14.0.1.106  3) python/2.7.9 => python/2.7.6
  2) mvapich2/1.9a2 => mvapich2/2.0b

[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 0, pid: 95159) exited with status 1
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][spawn_processes] Failed to execvp() 'c558-802':  (2)
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 1, pid: 95160) exited with status 1
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 2, pid: 95161) exited with status 1
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 3, pid: 95162) exited with status 1
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 4, pid: 95163) exited with status 1
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 5, pid: 95164) exited with status 1
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 6, pid: 95165) exited with status 1
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 7, pid: 95166) exited with status 1
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 8, pid: 95167) exited with status 1
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 9, pid: 95168) exited with status 1
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 10, pid: 95169) exited with status 1
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 11, pid: 95170) exited with status 1
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 12, pid: 95171) exited with status 1
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 13, pid: 95172) exited with status 1
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 14, pid: 95173) exited with status 1
[c558-802.stampede.tacc.utexas.edu:mpispawn_0][child_handler] MPI process (rank: 15, pid: 95174) exited with status 1
```

**Comment by marksantcroos** [comment]
Vivek, I'm looking into this.

One thing I noticed is, that you are performing computations in the pre_exec. The pre_exec is really just for things like settings environment_variables, etc, as these may not run on a compute node.

**Comment by marksantcroos** [comment]
It seems lsdmap has manouvred itself into a dependency nightmare.

The non-default `intel/14.0.1.106` pulls in a deprecated version of python (or at least warns about it, I dont actually see it loaded ... ) and a different version of mvapich2 than the agent uses.

You might be able to work around the latter by using a custom resource config that uses the same mvapich2.

I might be able to help you if you provide me with a minimal script with requires input data so that I can reproduce it, but I think the "general" fix would be to use the default versions of the libraries.
(And its not only the mpirun_rsh, but also the python-mpi executable that you use).

**Comment by marksantcroos** [comment]
I'm now wondering whether switching back to IBRUN would magically resolve this issues ... Will try that once I get the examples.

**Comment by vivek-bala** [comment]
Hey Mark,

These computations are pretty small and not time-consuming. If not the
pre_exec, the other method would be to have an exclusive CU for this
command (and then move the corresponding data between two or more CUs).
What do you suggest ?

One thing I noticed is, that you are performing computations in the
> pre_exec. The pre_exec is really just for things like settings
> environment_variables, etc, as these may not run on a compute node.
>
So you mean the pre_exec does not execute on the compute nodes ? In that
case, I understand running computations would be a violation.

-- Vivek

**Comment by vivek-bala** [comment]
>
> It seems lsdmap has manouvred itself into a dependency nightmare.
>
I agree. Took me a while to figure out these set of versions suitable to
run lsdmap.


> The non-default intel/14.0.1.106 pulls in a deprecated version of python
> (or at least warns about it, I dont actually see it loaded ... ) and a
> different version of mvapich2 than the agent uses.
>
I believe this loads python 2.7.6 (on intel/13* it is 2.7.9).


> You might be able to work around the latter by using a custom resource
> config that uses the same mvapich2.
>
You mean a custom resource config for extasy on stampede ?


> I might be able to help you if you provide me with a minimal script with
> requires input data so that I can reproduce it, but I think the "general"
> fix would be to use the default versions of the libraries.
> (And its not only the mpirun_rsh, but also the python-mpi executable that
> you use).
>
The python-mpi was suggested by a person from TACC as I was running into
mpi errors in lsdmap. PFA a folder with the slurm script and input data for
a lsdmap stage. I have used ibrun though. While launching the job, make
sure you have the default intel/13.0.2.146 module loaded (no gromacs module
with intel/14*).

**Comment by vivek-bala** [comment]
>
> I'm now wondering whether switching back to IBRUN would magically resolve
> this issues ... Will try that once I get the examples.
>
I think so, since earlier it was ibrun. Any specific reason why that was
changed to mpirun/mpirun_sh ?

**Comment by marksantcroos** [comment]
> PFA a folder with the slurm script and input data for a lsdmap stage.

Thanks, that helps!

> You mean a custom resource config for extasy on stampede ?

Possibly, but only as a last resort.

> I think so, since earlier it was ibrun. Any specific reason why that was changed to mpirun/mpirun_sh ?

Ok, then one of the options is to switch back to ibrun. We went one layer below because of overhead of ibrun. I don't feel strong about this so happy to use ibrun again, especially given that it is the official way of doing things.

Let me dig a bit further and then decide.

**Comment by marksantcroos** [comment]
So using IBRUN as LaunchMethod works indeed.

Changing the defaults (intel/14, etc.) for the Agent won't make much sense, because then Gromacs wouldn't work anymore.

What also works, is to not include the full pathname of `mpirun_rsh` during Agent discovery, so that it will pick up a "new" mpirun_rsh if that is implied by the pre_exec of the CU.

Based on my understanding of IBRUN, the reason why it works there, is for exactly this reason.

Any preference? Andre?

**Comment by vivek-bala** [comment]
I face similar mpi errors with CoCo as well. I verified that it works IBRUN (starts but fails because of a change in some of stampede's modules on april 28).

**Comment by andre-merzky** [comment]
detecting a new mpirun_rsh due to pre-exec seems to make most sense.  I would not be surprised if the same might eventually apply to other launch methods...

**Comment by marksantcroos** [comment]
Have created a pull request that resolves this issue.

**Comment by marksantcroos** [comment]
Hotfix release is in the making.

**Comment by marksantcroos** [comment]
Resolved in v0.32.

**Comment by marksantcroos** [comment]
Similar to #621 it seems ... will look into it.
